File: Symmetric Tree/Symmetric_Tree.cpp
// 第一种思路 使用递归
// 8ms 34.97%

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

class Solution {
public:
    bool isSymmetric(TreeNode* root) 
    {
        if (!root)
            return true;
        else
            return isMirror(root->left, root->right);
    }
    bool isMirror(TreeNode* left, TreeNode *right)
    {
        if (left == 0 && right == 0)
            return true;
        if (left == 0 || right == 0)
            return false;
        return (left->val == right->val) && isMirror(left->left, right->right) && isMirror(left->right, right->left);
    }
};

File: Symmetric Tree/Symmetric_Tree.py
# 第一种思路 使用递归
# 64ms 37.97%

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def isSymmetric(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """

        if root is None:
            return True
        else:
            return self.isMirror(root.left, root.right)
        
    def isMirror(self, left, right):
        if left is None and right is None:
            return True
        if left is None or right is None:
            return False
        return left.val == right.val and self.isMirror(left.left, right.right) and self.isMirror(left.right, right.left)